src/algebra/big_operators/norm_num.lean

```diff
@@ -0,0 +1,356 @@
+  pure (eis, eq)
+| `(@multiset.map %%α %%β %%ef %%exs) := do
+  (xs, xs_eq) ← tactic.norm_num.eval_multiset exs,
+  (ys, ys_eq) ← tactic.norm_num.eval_list_map ef xs,
+  eq ← i_to_expr ``(tactic.norm_num.multiset.map_congr %%ef %%xs_eq %%ys_eq),
+  pure (ys, eq)
+| e := fail (to_fmt "Unknown multiset expression" ++ format.line ++ to_fmt e)
+
+lemma list.cons_congr {α : Type*} (x : α) {xs : list α} {xs' : list α} (xs_eq : xs' = xs) :
+  x :: xs' = x :: xs :=
+by rw xs_eq
+
+lemma tactic.norm_num.list.map_congr {α β : Type*} (f : α → β) {xs xs' : list α}
+  {ys : list β} (xs_eq : xs = xs') (ys_eq : xs'.map f = ys) :
+  xs.map f = ys :=
+by rw [← ys_eq, xs_eq]
+
+/-- Convert an expression denoting a list to a list of elements. -/
+meta def tactic.norm_num.eval_list : expr → tactic (list expr × expr)
+| e@`(list.nil) := do
+  eq ← mk_eq_refl e,
+  pure ([], eq)
+| e@`(list.cons %%x %%xs) := do
+  (xs, xs_eq) ← tactic.norm_num.eval_list xs,
+  eq ← i_to_expr ``(list.cons_congr %%x %%xs_eq),
+  pure (x :: xs, eq)
+| e@`(list.range %%en) := do
+  n ← expr.to_nat en,
+  eis ← (list.range n).mmap (λ i, expr.of_nat `(ℕ) i),
+  eq ← mk_eq_refl e,
+  pure (eis, eq)
+| `(@list.map %%α %%β %%ef %%exs) := do
+  (xs, xs_eq) ← tactic.norm_num.eval_list exs,
+  (ys, ys_eq) ← tactic.norm_num.eval_list_map ef xs,
+  eq ← i_to_expr ``(tactic.norm_num.list.map_congr %%ef %%xs_eq %%ys_eq),
+  pure (ys, eq)
+| e := fail (to_fmt "Unknown list expression" ++ format.line ++ to_fmt e)
+
+@[to_additive]
+lemma list.prod_cons_congr {α : Type*} [monoid α] (xs : list α) (x y z : α)
+  (his : xs.prod = y) (hi : x * y = z) : (x :: xs).prod = z :=
+by rw [list.prod_cons, his, hi]
+
+/-- Evaluate `list.prod %%xs`,
+producing the evaluated expression and an equality proof. -/
+meta def list.prove_prod (α : expr) : list expr → tactic (expr × expr)
+| [] := do
+  result ← expr.of_nat α 1,
+  proof ← i_to_expr ``(@list.prod_nil %%α _),
+  pure (result, proof)
+| (x :: xs) := do
+  eval_xs ← list.prove_prod xs,
+  xxs ← i_to_expr ``(%%x * %%eval_xs.1),
+  eval_xxs ← or_refl_conv norm_num.derive xxs,
+  exs ← expr.of_list α xs,
+  proof ← i_to_expr
+    ``(list.prod_cons_congr %%exs%%x %%eval_xs.1 %%eval_xxs.1 %%eval_xs.2 %%eval_xxs.2),
+  pure (eval_xxs.1, proof)
+
+/-- Evaluate `list.sum %%xs`,
+sumucing the evaluated expression and an equality proof. -/
+meta def list.prove_sum (α : expr) : list expr → tactic (expr × expr)
+| [] := do
+  result ← expr.of_nat α 0,
+  proof ← i_to_expr ``(@list.sum_nil %%α _),
+  pure (result, proof)
+| (x :: xs) := do
+  eval_xs ← list.prove_sum xs,
+  xxs ← i_to_expr ``(%%x + %%eval_xs.1),
+  eval_xxs ← or_refl_conv norm_num.derive xxs,
+  exs ← expr.of_list α xs,
+  proof ← i_to_expr
+    ``(list.sum_cons_congr %%exs%%x %%eval_xs.1 %%eval_xxs.1 %%eval_xs.2 %%eval_xxs.2),
+  pure (eval_xxs.1, proof)
+
+@[to_additive] lemma list.prod_congr {α : Type*} [monoid α] {xs xs' : list α} {z : α}
+  (h₁ : xs = xs') (h₂ : xs'.prod = z) : xs.prod = z := by cc
```

Comment:
This type of thing doesn't feel like it should be in the root namespace, anyone who thinks they want this lemma for anything other than metaprogramming probably wants `rwa h₁` or similar instead.

Reply:
This is now fixed right?

Reply:
Yes, everything is in the `tactic.norm_num` namespace (line 36).